File: c/bowling/src/bowling.c
int score(const int rolls[]) {
    int total = 0;
    int roll_index = 0;
    
    for (int frame = 0; frame < 10; frame++) {
        if (rolls[roll_index] + rolls[roll_index + 1] == 10) {
            // Spare: add 10 plus the next roll as bonus
            total += 10 + rolls[roll_index + 2];
            roll_index += 2;
        } else {
            // Normal frame: just add the two rolls
            total += rolls[roll_index] + rolls[roll_index + 1];
            roll_index += 2;
        }
    }
    
    return total;
}

File: go/bowling/bowling.go
package main

type Game struct {
	Rolls []int
}

func NewGame() Game {
	game := Game{
		Rolls: make([]int, 20),
	}
	return game
}

func (game Game) Roll(pins int) Game {
	game.Rolls = append(game.Rolls, pins)
	return game
}

func (game Game) Score() int {
	return doScore(game.Rolls)
}

func doScore(pins []int) int {
	if len(pins) <= 3 {
		return sum(pins)
	}
	x, y, z := pins[0], pins[1], pins[2]
	if x == 10 {
		return x + y + z + doScore(pins[1:])
	} else if x+y == 10 {
		return x + y + z + doScore(pins[2:])
	} else {
		return x + y + doScore(pins[2:])
	}
}

func sum(pins []int) int {
	score := 0
	for p := range len(pins) {
		score += pins[p]
	}
	return score
}

func main() {}
